datastore.py

```diff
@@ -0,0 +1,38 @@
+import boto3
+import json
+
+from botocore.errorfactory import ClientError
+
+s3 = boto3.resource('s3')
+
+class Datastore():
+    """Datastore allows us to persist some data outside the application.
+    
+    The data property is loaded initially from S3 and then cached for the
+    lifetime of the application (subsequent reads will not pull from S3).
+    """
+    def __init__(self, s3_bucket="porkbutts-discoball", s3_key="data.json"):
+        self.s3_obj = s3.Object(s3_bucket, s3_key)
+        self.__data = None
+        
+    @property
+    def data(self):
+        """Return data if we have it cached otherwise retrieve from S3.
+
+        Returns:
+            object: arbitrary json data deserialized to a python dict
+        """
+        if not self.__data:
+            try:
+                contents = self.s3_obj.get()['Body'].read().decode('utf-8')
+                self.__data = json.loads(contents)
+            except ClientError:
+                self.__data = {}
+        return self.__data
+     
+     
+    def save(self):
+        """Serializes self.__data as json and saves the result to S3.
+        """
+        self.s3_obj.put(Body=json.dumps(self.__data))

```

Comment:
From what I gather, the `Body` is a StreamingBody() type which is why earlier (line 27) you needed `read()` to turn it into byte and `decode()` it into utf-8. When you `put()`, do you not need to encode the json string into bytes and turn it into StreamingBody(), or does aws do all that for you?

Reply:
I just copied that the load from stackoverflow. i dont think utf-8 is necessary actually. `read()` will basically take all the data from the stream and put it into a byte buffer. decode will turn the byte buffer into a string. this is cuz the file could be binary data, we dont know and s3 isn’t gonna convert it to whatever type for you.

actually looking on SO, `json.load` which usually takes a file seems like it can take a streaming body.

Reply:
That still doesn't answer my question of why we don't need to put `json.dumps(self.__data)` string into a StreamingBody type. Pulling up the documentation of put(), looks like it takes `Body=b'bytes'|file` as an argument. so a json string is a file?

Reply:
json.dumps returns a string, so it’s more likely that the string is somehow getting casted to bytes not file.